core/bootstrap.go

```diff
@@ -16,108 +17,187 @@ import (
+	// this threshold can afford to be small (<=30s).
+	BootstrapPeriod = 30 * time.Second
+
+	// BootstrapPeerThreshold governs the node Bootstrap process. If the node
+	// has less open connections than this number, it will open connections
+	// to the bootstrap nodes. From there, the routing system should be able
+	// to use the connections to the bootstrap nodes to connect to even more
+	// peers. Routing systems like the IpfsDHT do so in their own Bootstrap
+	// process, which issues random queries to find more peers.
+	BootstrapPeerThreshold = 4
+
+	// BootstrapConnectionTimeout determines how long to wait for a bootstrap
+	// connection attempt before cancelling it.
+	BootstrapConnectionTimeout time.Duration = BootstrapPeriod / 3
 )
 
-func superviseConnections(parent context.Context,
-	h host.Host,
-	route *dht.IpfsDHT, // TODO depend on abstract interface for testing purposes
-	store peer.Peerstore,
-	peers []peer.PeerInfo) error {
+// nodeBootstrapper is a small object used to bootstrap an IpfsNode.
+type nodeBootstrapper struct {
+	node *IpfsNode
+}
 
-	for {
-		ctx, _ := context.WithTimeout(parent, connectiontimeout)
-		// TODO get config from disk so |peers| always reflects the latest
-		// information
-		if err := bootstrap(ctx, h, route, store, peers); err != nil {
+// TryToBootstrap starts IpfsNode bootstrapping. This function will run an
+// initial bootstrapping phase before exiting: connect to several bootstrap
+// nodes. This allows callers to call this function synchronously to:
+// - check if an error occurrs (bootstrapping unsuccessful)
+// - wait before starting services which require the node to be bootstrapped
+//
+// If bootstrapping initially fails, Bootstrap() will try again for a total of
+// three times, before giving up completely. Note that in environments where a
+// node may be initialized offline, as normal operation, BootstrapForever()
+// should be used instead.
+//
+// Note: this function could be much cleaner if we were to relax the constraint
+// that we want to exit **after** we have performed initial bootstrapping (and are
+// thus connected to nodes). The constraint may not be that useful in practice.
+// Consider cases when we initialize the node while disconnected from the internet.
+// We don't want this launch to fail... want to continue launching the node, hoping
+// that bootstrapping will work in the future if we get connected.
+func (nb *nodeBootstrapper) TryToBootstrap(ctx context.Context, peers []peer.PeerInfo) error {
+	n := nb.node
+
+	// TODO what bootstrapping should happen if there is no DHT? i.e. we could
+	// continue connecting to our bootstrap peers, but for what purpose? for now
+	// simply exit without connecting to any of them. When we introduce another
+	// routing system that uses bootstrap peers we can change this.
+	dht, ok := n.Routing.(*dht.IpfsDHT)
+	if !ok {
+		return nil
+	}
+
+	for i := 0; i < 3; i++ {
```

Comment:
3 is `3*bootstrap peers`. it's not clear what to do in case this fails. `BootstrapForever` has issues too (makes some tests fail). But agree that the daemon should be able to start if bootstrap fails (which i just realized wont work in this case).

It may be best to forgo the two constraints mentioned there:
```
// - check if an error occurrs (bootstrapping unsuccessful)
// - wait before starting services which require the node to be bootstrapped
```